teps/0022-trigger-immutable-input.md

```diff
@@ -0,0 +1,351 @@
+type. Incoming event data will be stored within `input` as `body` and `headers`
+fields.
+
+##### Example
+
+Before:
+
+```
+POST /interceptor HTTP/1.1
+Host: my-interceptor:4567
+X-GitHub-Delivery: 72d3162e-cc78-11e3-81ab-4c9367dc0958
+X-Hub-Signature: sha1=7d38cdd689735b008b3c702edd92eea23791c5f6
+User-Agent: GitHub-Hookshot/044aadd
+Content-Type: application/json
+Content-Length: 6615
+X-GitHub-Event: issues
+Tekton-Event-ID: abc123
+{
+  "action": "opened",
+  "issue": {
+    "url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
+    "number": 1347,
+    ...
+  },
+  "repository" : {
+    "id": 1296269,
+    "full_name": "octocat/Hello-World",
+    "owner": {
+      "login": "octocat",
+      "id": 1,
+      ...
+    },
+    ...
+  },
+  "sender": {
+    "login": "octocat",
+    "id": 1,
+    ...
+  }
+}
+```
+
+After:
+
+```
+POST /interceptor HTTP/1.1
+Host: my-interceptor:4567
+Tekton-Event-ID: abc123
+{
+  "input": {
+    "headers": [
+      "X-GitHub-Delivery": "72d3162e-cc78-11e3-81ab-4c9367dc0958",
+      "X-Hub-Signature": "sha1=7d38cdd689735b008b3c702edd92eea23791c5f6",
+      "User-Agent": "GitHub-Hookshot/044aadd",
+      "Content-Type": "application/json",
+      "Content-Length": "6615",
+      "X-GitHub-Event": "issues",
+    ],
+    "body": {
+      "action": "opened",
+      "issue": {
+        "url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
+        "number": 1347,
+        ...
+      },
+      "repository" : {
+        "id": 1296269,
+        "full_name": "octocat/Hello-World",
+        "owner": {
+          "login": "octocat",
+          "id": 1,
+          ...
+        },
+        ...
+      },
+      "sender": {
+        "login": "octocat",
+        "id": 1,
+        ...
+      }
+    }
+  }
+}
+```
+
+#### Interceptor Field
+
+We propose that any responses returned from interceptors are now additive to the
+interceptor chain payload, and remove the ability to arbitrarily modify
+payloads.
+
+Any response received from interceptors will be included in following paylods
+under a new `interceptors` field, keyed by interceptor name for uniqueness.
+Webhook interceptors may no longer modify headers as part of the interceptor
+chain.
+
+If no response body is returned by the interceptor, no `interceptors.<name>`
+field will be present.
+
+##### Example
+
+Given a webhook interceptor named `foo` that returns the following response:
+
+```json
+{
+  "bar": "baz"
+}
+```
+
+We expect the next request body to the following interceptor to look like:
+
+```json
+{
+  "input": {...},
+  "interceptors": {
+    "foo": {
+      "bar": "baz"
+    }
+  }
+}
+```
+
+### CEL Interceptors and TriggerBindings
+
+To stay consistent with Webhook inteceptors, we will now key input event fields
+for CEL Interceptors and TriggerBindings on `input` instead of `body` and
+`header` - e.g. `body` => `input.body` and `header` => `input.header`.
```

Comment:
Besides consistency, is there a disadvantage to keeping just `body` and `header` instead of `input.body` and `input.header`. Maybe for webhooks too, we can just send `header` and `body` as top-level fields?

The main benefit is that then this change becomes a lot less invasive i.e. users do not have to rewrite their bindings/CEL interceptors (only their webhook ones).